### PR TITLE
Allow List Handling

### PR DESCRIPTION
**Pull request by lgostling**
…custom nodes

### CODE REVIEW

File: Core.Collectors/Collector/CollectorBase.cs [hunk 1]
```diff
@@ -63,8 +63,18 @@ public async Task<bool> ProcessAsync(T collectionNode, long maxPageCount)
             while (batchingHttpRequest.HasNext && counter < maxPageCount && !haltCollection)
             {
                 counter++;
-                HttpResponseMessage response = await batchingHttpRequest.NextResponseAsync(this.authentication).ConfigureAwait(false);
+                RequestResult result = await batchingHttpRequest.NextResponseAsync(this.authentication).ConfigureAwait(false);
 
+                if (!result.IsSuccess())
+                {
+                    foreach(T allowListCollectionNode in result.allowListStatus.Continuation())
```
Comment (kivancmuslu) [inline Core.Collectors/Collector/CollectorBase.cs:70]: >T [](http://example.com/codeflow?start=28&length=1)

Are we certain that this cast is fine? Continuation() here returns a CollectionNode, so in theory, this could fail, no?

Comment (kivancmuslu) [inline Core.Collectors/Collector/CollectorBase.cs:70]: >h [](http://example.com/codeflow?start=26&length=1)

Nit: space after foreach

File: Core.Collectors/Collector/CollectorBase.cs [hunk 2]
```diff
@@ -63,8 +63,18 @@ public async Task<bool> ProcessAsync(T collectionNode, long maxPageCount)
...
+                    {
+                        await this.ProcessAsync(allowListCollectionNode).ConfigureAwait(false);
+                    }
+                    return false;
```
Comment (kivancmuslu) [inline Core.Collectors/Collector/CollectorBase.cs:74]: >return false; [](http://example.com/codeflow?start=20&length=13)

Why not "break"?

File: Core.Collectors/Web/HttpExceptionSignature.cs [hunk 3]
```diff
@@ -1,29 +1,49 @@
...
 using System.Threading.Tasks;
 
 namespace Microsoft.CloudMine.Core.Collectors.Web
-{
-    public class HttpExceptionSignature
+{ 
+    public class HttpExceptionSignature : IAllowListStatus
     {
-        public static readonly HttpExceptionSignature RequestTimeoutException = new HttpExceptionSignature(exception => 
+        public static HttpExceptionSignature RequestTimeoutException(Func<List<CollectionNode>> continuation = null)
```
Comment (kivancmuslu) [inline Core.Collectors/Web/HttpExceptionSignature.cs:13]: >(Func<List<CollectionNode>> continuation = null) [](http://example.com/codeflow?start=68&length=48)

It looks like that making this a method (instead of a property) is causing quite a bit of changes. Can you do this in a separate PR or just skip doing this?

File: Core.Collectors/Web/HttpExceptionSignature.cs [hunk 4]
```diff
@@ -1,29 +1,49 @@
...
         }
 
         public bool Matches(Exception exception)
         {
             return matcher.Invoke(exception);
         }
+
+        public List<CollectionNode> Continuation()
+        {
+            return this.continuation.Invoke();
```
Comment (kivancmuslu) [inline Core.Collectors/Web/HttpExceptionSignature.cs:46]: >return this.continuation.Invoke(); [](http://example.com/codeflow?start=12&length=34)

Nit: in general, I prefer seeing:

```
this.continuation();
```
(rather than explicit Invoke)

File: Core.Collectors/Web/HttpResponseSignature.cs [hunk 5]
```diff
@@ -1,29 +1,39 @@
...
 
 namespace Microsoft.CloudMine.Core.Collectors.Web
 {
-    public class HttpResponseSignature
+    public class HttpResponseSignature : IAllowListStatus
     {
         private readonly HttpStatusCode statusCode;
         private readonly Regex responseMessageRegex;
         private readonly Func<JObject, bool> matcher;
+        private readonly List<CollectionNode> continuation;
```
Comment (kivancmuslu) [inline Core.Collectors/Web/HttpResponseSignature.cs:18]: Why don't we also make this a function-based continuation (just in case we need recursion here as well)? That would also keep things similar.

File: Core.Collectors/Web/IAllowListStatus.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CloudMine.Core.Collectors.Web
+{
+    public interface  IAllowListStatus
+    {
+
+        public List<CollectionNode> Continuation();
```
Comment (kivancmuslu) [inline Core.Collectors/Web/IAllowListStatus.cs:11]: Nit: empty line.

File: Core.Collectors/Web/IAllowListStatus.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.CloudMine.Core.Collectors.Collector;
+using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (kivancmuslu) [inline Core.Collectors/Web/IAllowListStatus.cs:4]: Do you need all of these?

File: Core.Collectors/Web/RequestResult.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+using System.Collections.Generic;
+using System.Net.Http;
+
+namespace Microsoft.CloudMine.Core.Collectors.Web
+{
+    public class RequestResult
+    {
+        public HttpResponseMessage response;
+        //public Func<List<CollectionNode>> continuation;
+        public IAllowListStatus allowListStatus;
```
Comment (kivancmuslu) [inline Core.Collectors/Web/RequestResult.cs:12]: Remove.

File: Core.Collectors/Web/RequestResult.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
+﻿using Microsoft.CloudMine.Core.Collectors.Collector;
+using System;
+using System.Collections.Generic;
```
Comment (kivancmuslu) [inline Core.Collectors/Web/RequestResult.cs:3]: Do you need this?

**Review by kivancmuslu — CHANGES_REQUESTED** [review]
:clock1:

File: Core.Collectors/Web/HttpExceptionSignature.cs (hunk 4 above)
Comment (AllainPL) [reply Core.Collectors/Web/HttpExceptionSignature.cs:46]: trivia:
I will add on this, it's not only the matter of preference it actually "may" make a perf difference.
`.Invoke()` has the benefit of not caring about anything in terms of typing - to do this - it needs to do much more during execution instead of compile time. Invoke operates on `Object`s and `Binder`s and can execute "anything" out of thin air, while a regular call has the benefit of being generated properly by the compiler with all the necessary code but it requires all the details about the function to be known at compilation time.

File: Core.Collectors/Collector/CollectorBase.cs (hunk 1 above)
Comment (lgostling) [reply Core.Collectors/Collector/CollectorBase.cs]: Address this in new commit and opened a new comment.

File: Core.Collectors/Collector/CollectorBase.cs [hunk 10]
```diff
@@ -63,8 +63,21 @@ public async Task<bool> ProcessAsync(T collectionNode, long maxPageCount)
...
             {
                 counter++;
-                HttpResponseMessage response = await batchingHttpRequest.NextResponseAsync(this.authentication).ConfigureAwait(false);
+                RequestResult result = await batchingHttpRequest.NextResponseAsync(this.authentication).ConfigureAwait(false);
 
+                if (!result.IsSuccess())
+                {
+                    foreach (CollectionNode allowListCollectionNode in result.allowListStatus.Continuation())
+                    {
+                        if (allowListCollectionNode is T)
```
Comment (lgostling) [inline Core.Collectors/Collector/CollectorBase.cs:72]: @kivancmuslu Is this a good solution for casting Collection node to T?

File: Core.Collectors/Collector/CollectorBase.cs [hunk 11]
```diff
@@ -63,8 +63,21 @@ public async Task<bool> ProcessAsync(T collectionNode, long maxPageCount)
...
+                        {
+                            await this.ProcessAsync((T) allowListCollectionNode).ConfigureAwait(false);
+                        }
```
Comment (kivancmuslu) [inline Core.Collectors/Collector/CollectorBase.cs:75]: There might be a better way to auto-cast these stuff, but as long as this is working, I am good for letting this in as it is. I will see if I can make generics happy later on :)

File: Core.Collectors/Collector/CollectorBase.cs (hunk 10 above)
Comment (kivancmuslu) [reply Core.Collectors/Collector/CollectorBase.cs:72]: There might be a better way to auto-cast these stuff, but as long as this is working, I am good for letting this in as it is. I will see if I can make generics happy later on :)

(Resolve this comment after you read it)

File: Core.Collectors/Web/HttpExceptionSignature.cs [hunk 12]
```diff
@@ -1,29 +1,42 @@
...
         }
 
         public bool Matches(Exception exception)
         {
             return matcher.Invoke(exception);
         }
+
+        public List<CollectionNode> Continuation()
+        {
+            if(continuation == null)
```
Comment (kivancmuslu) [inline Core.Collectors/Web/HttpExceptionSignature.cs:34]: > if [](http://example.com/codeflow?start=11&length=3)

Nit: space after 'if'

**Comment by kivancmuslu** [comment]
> using System.Net.Http;

Nit: missing copyright.

---
Refers to: Core.Collectors/Web/RequestResult.cs:1 in f875a8b. [](commit_id = f875a8b6f7ad34025f10712835756439254ce2ed, deletion_comment = False)

**Comment by kivancmuslu** [comment]
As long as the if check and internal cast against T is working in the code (please ensure that the path is exercise in the debugger), I am good with the change as it is. I will take a look at it later to see if I can make generics happier. Out of my head, I don't also know the best way to represent this.

**Review by kivancmuslu — APPROVED** [review]
:shipit:
